Fix apache#47555: PostgreSQL URI driver issue

Dong-yeong0 opened the pull request:
This PR fixes an issue(#47555) where Airflow incorrectly generates a PostgreSQL connection URI using `postgres://` instead of `postgresql://`, which causes SQLAlchemy to fail.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Dong-yeong0 committed:
Fix apache#47555: PostgreSQL URI scheme issue

Dong-yeong0 committed:
Fix apache#47555: PostgreSQL URI scheme issue

andresitorusz commented on airflow/models/connection.py:
```diff
@@ -256,6 +256,8 @@ def get_uri(self) -> str:
         if self.conn_type:
             uri = f"{self.conn_type.lower().replace('_', '-')}://"
+            if self.conn_type == "postgres":
+                uri = uri.replace("postgres://", "postgresql://", 1)
```
I think it's better to pass `postgres` or `postgresql` as a parameter (maybe adding an extra field in the connection form) instead of changing it for every postgres connection.

Based on this [documentation](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING-URIS), both should work.

Dong-yeong0 replied:
Thank you for your feedback 👍👍 

I appreciate your idea, and I've updated the code (9e1a54d29fc89ce96ebc1c1d5a1170ab88227ae4) based on your suggestion.

Dong-yeong0 committed:
Fix: allow configuring PostgreSQL connection driver

Added support for customizing the PostgreSQL connection driver (`postgres` vs `postgresql`)
via the `postgres_driver` option in the extra JSON field.

Dong-yeong0 committed:
Fix: allow configuring PostgreSQL connection driver

Added support for customizing the PostgreSQL connection driver (`postgres` vs `postgresql`)
via the `postgres_driver` option in the extra JSON field.

Dong-yeong0 committed:
Merge branch 'fix-postgres-uri' of https://github.com/Dong-yeong0/airflow into fix-postgres-uri

andresitorusz commented on airflow/models/connection.py:
```diff
@@ -256,6 +256,9 @@ def get_uri(self) -> str:
             uri = f"{self.conn_type.lower().replace('_', '-')}://"
+            if self.conn_type == "postgres":
+                driver = self.extra_dejson.get("postgres_driver", "postgres")
+                uri = uri.replace("postgres://", f"{driver}://", 1)
```
```suggestion
            if self.conn_type == "postgres":
                designator = self.extra_dejson.get("designator", "postgres")
                uri = uri.replace("postgres://", f"{designator}://", 1)
```

andresitorusz replied:
Also, please add the `designator` documentation in the [PostgreSQL Extra](https://github.com/apache/airflow/blob/main/providers/postgres/docs/connections/postgres.rst?plain=1#L49) documentation.

Dong-yeong0 replied:
I appreciate your suggestion 👍

It helped me learn something new, and I have updated (876ea6f8f432220f6c46b042b3b281b36ea13661) the documentation accordingly.

andresitorusz replied:
Great, I'm also still learning. Kindly check the suggested change in the above comment as well.

andresitorusz replied:
To add more, actually https://github.com/apache/airflow/pull/38831 solves your problem. But I think it's also good if we can customize the `designator` via the extra bcoz I also faced the issue back then when connecting to Postgres using JDBC.

Dong-yeong0 replied:
Thank you for pointing out #38831 👍

I see that it addresses the `get_uri()` issue.

I'll review its changes and check if customizing the `designator` via `extra` still adds value.

If needed, I'll adjust my implementation accordingly.

Thanks again for your insights! 🚀

Dong-yeong0 committed:
Add the documentation for the 'designator' parameter in the PostgreSQL extras section

Dong-yeong0 committed:
Merge branch 'main' into fix-postgres-uri

jason810496 commented on airflow/models/connection.py:
```diff
@@ -256,6 +256,9 @@ def get_uri(self) -> str:
 
         if self.conn_type:
             uri = f"{self.conn_type.lower().replace('_', '-')}://"
+            if self.conn_type == "postgres":
```
Could you add a test for this change?  
The test is located in `tests/models/test_connection.py::TestConnection::test_get_uri`.

Dong-yeong0 replied:
After reading potiuk's comment below, I realized that I was completely mistaken.

I really appreciate your suggestion👍, but I think I need to change my approach.

jason810496 replied:
Agree with potiuk's comment.
> I really appreciate your suggestion👍, but I think I need to change my approach.

Sure, look forward for your change, thanks !

jason810496 reviewed (commented):
Looks good! Only the test needs to be added.

potiuk reviewed (changes requested):
I don't thnk this is needed/useful. Get_uri - returns "airflow" URI format as described in https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#uri-format-example, the #38831 was indeed a way how to get `another` uri (sqlalchemy URI) - which is pretty different than AIrfow's has different potential parameters and uses `postgres` as prefix exclusively.

If anything - maybe it's worth (and change this PR into it) to add comment that this is NOT sqlalchemy URI and that `sqlalchemy_url` should be used instead. 

It also makes it very fishy to add postgres-specific code in airflow models anyway.

Dong-yeong0 committed:
Remove the code that constructs the URI based on connection type

Dong-yeong0 committed:
Update get_uri method docstring

Dong-yeong0 committed:
Clarify that get_uri() does not return a SQLAlchemy URI

Dong-yeong0 committed:
Merge branch 'main' into fix-postgres-uri

Dong-yeong0 committed:
Fix rst-backticks

Dong-yeong0 committed:
Merge branch 'main' into fix-postgres-uri

andresitorusz commented on providers/postgres/docs/connections/postgres.rst:
```diff
@@ -76,6 +76,8 @@ Extra (optional)
       For more information, refer to the psycopg2 documentation on `connection and cursor subclasses <https://www.psycopg.org/docs/extras.html#connection-and-cursor-subclasses>`_.
 
+    * ``designator`` - Specifies the designator to use in the connection URI. By default, it is set to ``postgres``. This is useful if you want to replace the ``postgres://`` prefix in the URI with a different designator.
+
```
You should also delete this one because we are not to implement the previous approach.

Dong-yeong0 replied:
Okay, i'll remove it

Dong-yeong0 committed:
Remove designer description

potiuk reviewed (approved):
NICE

jason810496 commented on airflow/models/connection.py:
```diff
@@ -26,14 +26,14 @@
 from typing import Any
 from urllib.parse import parse_qsl, quote, unquote, urlencode, urlsplit
 
+from airflow.sdk.execution_time.secrets_masker import mask_secret
```
You could run the following `pre-commit` and commit to fix.
```
pre-commit run ruff --all && pre-commit run ruff-format --all 
```

jason810496 reviewed (changes requested):
Thanks @Dong-yeong0, but the `ruff` static check need fix.

Dong-yeong0 committed:
Fix format code with ruff

Dong-yeong0 committed:
Merge branch 'main' into fix-postgres-uri

jason810496 committed:
Merge branch 'main' into fix-postgres-uri

jason810496 committed:
Merge branch 'main' into fix-postgres-uri
